src/FrozenBubble/CannonBubble.java

```diff
@@ -0,0 +1,252 @@
+                canvas.remove(this);
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * @return the color of the cannon bubble.
+     */
+    public Color getColor() {
+        return color;
+    }
+
+    /**
+     * @return the x velocity of the cannon bubble.
+     */
+    public double getXVelocity() {
+        return xVelocity;
+    }
+
+    /**
+     * @return the x position of the cannon bubble.
+     */
+    public double getXPosition() {
+        return xPosition;
+    }
+
+    /**
+     * @return the y position of the cannon bubble
+     */
+    public double getYPosition() {
+        return yPosition;
+    }
+
+    /**
+     * @return the x position of the center of the cannon bubble
+     */
+    public double getCenterX() {
+        return xPosition + width / 2;
+    }
+
+    /**
+     * @return the y position of the center of the cannon bubble
+     */
+    public double getCenterY() {
+        return yPosition + height / 2;
+    }
+
+    /**
+     * Find the neighbour bubbles of the cannon bubble.
+     * 
+     * @param canvas the canvas where the cannon bubble at
+     * @return the set of the neighbour bubbles of the cannon bubble
+     */
+    public Set<Bubble> getNeighbours(CanvasWindow canvas) {
+
+        Set<Bubble> neighbourBubbles = new HashSet<>();
+        double centerX = getCenterX();
+        double centerY = getCenterY();
+
+        // add bubble1(the one at the top-left) to the set
+        if (canvas.getElementAt(centerX - width / 2, centerY - height) != null) {
+            neighbourBubbles.add((Bubble) canvas.getElementAt(centerX - width / 2, centerY - height));
+        }
+
+        // add bubble2(the one at the top-right) to the set
+        if (canvas.getElementAt(centerX + width / 2, centerY - height) != null) {
+            neighbourBubbles.add((Bubble) canvas.getElementAt(centerX + width / 2, centerY - height));
+        }
+
+        // add bubble3(the one at the most left) to the set
+        if (canvas.getElementAt(centerX - width, centerY) != null) {
+            neighbourBubbles.add((Bubble) canvas.getElementAt(centerX - width, centerY));
+        }
+
+        // add bubble4(the one at the most right) to the set
+        if (canvas.getElementAt(centerX + width, centerY) != null) {
+            neighbourBubbles.add((Bubble) canvas.getElementAt(centerX + width, centerY));
+        }
+
+        // add bubble5(the one at the bottom-left) to the set
+        if (canvas.getElementAt(centerX - width / 2, centerY + height) != null) {
+            neighbourBubbles.add((Bubble) canvas.getElementAt(centerX - width / 2, centerY + height));
+        }
+
+        // add bubble6(the one at the bottom-right) to the set
+        if (canvas.getElementAt(centerX + width / 2, centerY + height) != null) {
+            neighbourBubbles.add((Bubble) canvas.getElementAt(centerX + width / 2, centerY + height));
+        }
+
+        return neighbourBubbles;
+    }
```

Comment:
This _large_ amount of repeated code definitely suggests `Bubble` and `CannonBubble` should not be separate classes.